### PR TITLE
Updated SiStrip unpacker (new FED readout modes)

### PR DESCRIPTION
**Pull request by forthommel**
Introducing changes from #12157 and #12385 (from 7_5_X) to the 8_0_X release to be able to unpack new data formats introduced in the last FED FW upgrade (see these two PRs for more details)

### CODE REVIEW

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
apparently, this one still merges OK in 80X after #12307 is merged.
Nice.

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc [hunk 1]
```diff
@@ -355,9 +355,28 @@ try {
...
 	}
-      } else {
+    } else if likely(mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8 ) {
+
+      try {
+        // create unpacker
+        sistrip::FEDZSChannelUnpacker unpacker = sistrip::FEDZSChannelUnpacker::zeroSuppressedLiteModeUnpacker(buffer->channel(fedCh));
+
+        // unpack
+        clusterizer.addFed(unpacker,ipair,record);
```
Comment (VinInn) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: this line wil not compile anymore...

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc [hunk 2]
```diff
@@ -355,9 +355,28 @@ try {
...
+            ss << "Unordered clusters for channel " << fedCh << " on FED " << fedId << ": " << e.what();
+            edm::LogWarning(sistrip::mlRawToCluster_) << ss.str();
+          }
+          continue;
+        }
+
+    } else {
       
-      if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED ) { 
+      if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED or mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_FAKE ) {
```
Comment (VinInn) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: please use "||" nor "or"

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc [hunk 3]
```diff
@@ -355,9 +355,28 @@ try {
 	  }                                               
 	  continue;
 	}
-      } else {
+    } else if likely(mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8 ) {
```
Comment (VinInn) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: do not see any difference w/r/t the previous case (LITE10)
please collapse the two

Comment (forthommel) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: I left these two separated as I wanted to start using the static ZS lite unpacker defined in the SiStrip BSUnpacker (BS = bit stripping), which can be used for basically any word ADC length (here, 8-bit). Unfortunately the time constraints prevented me to test it so far...
So I guess I can collapse these as from now.

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc (hunk 1 above)
Comment (forthommel) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: I cannot reproduce on my test bench. Could you please give me your compiler output?

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc (hunk 2 above)
Comment (forthommel) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: Sure. Will do in the next commit!

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc (hunk 1 above)
Comment (VinInn) [inline RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: you need to merge PR  #12307 
if you collapse Lite8 and lite10 there will be no problem..

**Comment by slava77** [comment]
@forthommel 
please remove the merge of 12307, it's already in the 80X branch.
Either rebase to CMSSW_8_0_X_2015-11-12-1100 or remake in this IB from scratch.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12387	 d084963
- code changes look consistent with 75X/76X modifications; although the cleanup made in the processing options as seen in this PR and not available in 76X/75X (see below) could have been propagated back to 76X/75X
- jenkins tests pass and comparisons with the baseline show no differences


comparison of "76X version" .. "80X version (this PR)"
```diff
--- a/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc
+++ b/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc
@@ -346,7 +346,7 @@ try { // edmNew::CapacityExaustedException
     const sistrip::FEDReadoutMode mode = buffer->readoutMode();
     
     
-    if likely(mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE10 ) { 
+    if likely(mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE10 || mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8) { 
        
        try {
          // create unpacker
@@ -370,28 +370,9 @@ try { // edmNew::CapacityExaustedException
          }                                               
          continue;
        }
-    } else if likely(mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE8 ) {
-
-      try {
-        // create unpacker
-        sistrip::FEDZSChannelUnpacker unpacker = sistrip::FEDZSChannelUnpacker::zeroSuppressedLiteModeUnpacker(buffer->channel(fedCh));
-
-        // unpack
-        clusterizer.addFed(unpacker,ipair,record);
-      } catch (edmNew::CapacityExaustedException) {
-        throw;
-      } catch (const cms::Exception& e) {
-          if (edm::isDebugEnabled()) {
-            std::ostringstream ss;
-            ss << "Unordered clusters for channel " << fedCh << " on FED " << fedId << ": " << e.what();
-            edm::LogWarning(sistrip::mlRawToCluster_) << ss.str();
-          }
-          continue;
-        }
-
     } else {
       
-      if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED or mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_FAKE ) { 
+      if (mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED || mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED_FAKE ) { 
        try {
          // create unpacker
          sistrip::FEDZSChannelUnpacker unpacker = sistrip::FEDZSChannelUnpacker::zeroSuppressedModeUnpacker(buffer->channel(fedCh));
```

**Comment by davidlange6** [comment]
+1

**Comment by Martin-Grunewald** [comment]
Can you please clarify the value to be used for LegacyUnpacker:
For Run-1 data it is TRUE, for Run-2 data it is FALASE?
Or is it true for all data taken up to now, and false only for future 2016+ data?
How is it the case with MC?

**Comment by slava77** [comment]
IIRC, LegacyUnpacker=true is introduced mostly for safety. The value "false" so far appeared to work OK for all.
@forthommel should confirm.

**Pull request merged**